File: 백준/Silver/2805. 나무 자르기/나무 자르기.py
N, M = map(int, input().split())
trees = list(map(int, input().split()))
low = 1
high = max(trees)

ans = 0
while low <= high:
    mid = (high+low) // 2

    able = 0
    for tree in trees:
        if tree > mid:
            able += tree - mid

    if able == M:
        ans = mid
        break
    else:
        if able > M:
            low = mid + 1
            if ans < mid:
                ans = mid
        else:
            high = mid - 1
print(ans)

File: 백준/Silver/2805. 나무 자르기/나무 자르기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;
public class Main {
	public static void main(String[] args) throws IOException {
		
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine(), " ");
		int N = Integer.parseInt(st.nextToken());
		int M = Integer.parseInt(st.nextToken());
		int[] trees = new int[N];
		int low = 0;
		int high = 0;
		
		st = new StringTokenizer(br.readLine(), " ");
		for(int i = 0; i < N; i++) {
			trees[i] = Integer.parseInt(st.nextToken());

			if(high < trees[i]) {
				high = trees[i];
			}
		}
		
		while(low < high) {
			
			int mid = (low + high) / 2;
			long able = 0;
			for(int tree : trees) {

				if(tree - mid > 0) { 
					able += (tree - mid);
				}
			}

			if(able < M) {
				high = mid;
			}

			else {
				low = mid + 1;
			}
		}
		System.out.println(low - 1);
	}
}